src/prefect/_internal/concurrency/supervisors.py

```diff
@@ -119,65 +88,97 @@ def from_call(cls, __fn, *args, **kwargs) -> "WorkItem":
             context=contextvars.copy_context(),
         )
 
-    def __post_init__(self):
-        logger.debug("Created work item %r", self)
-
-    def run(self):
+    def run(self) -> None:
         """
-        Execute the work item.
+        Execute the call and place the result on the future.
 
-        All exceptions during execution of the work item are captured and attached to
-        the future.
+        All exceptions during execution of the call are captured.
         """
         # Do not execute if the future is cancelled
         if not self.future.set_running_or_notify_cancel():
+            logger.debug("Skipping execution of cancelled callback %s", self)
             return
 
-        logger.debug("Running work item %s", self)
+        logger.debug("Running callback %s", self)
 
-        # Execute the work and set the result on the future
-        if inspect.iscoroutinefunction(self.fn):
+        coro = self._run_sync()
+        if coro is not None:
             if get_running_loop():
                 # If an event loop is available, return a coroutine to be awaited
-                return self._run_async()
+                return self._run_async(coro)
             else:
                 # Otherwise, execute the function here
-                return asyncio.run(self._run_async())
-        else:
-            return self._run_sync()
+                return asyncio.run(self._run_async(coro))
+
+        return None
 
     def _run_sync(self):
         try:
             result = self.context.run(self.fn, *self.args, **self.kwargs)
+
+            # Return the coroutine for async execution
+            if inspect.isawaitable(result):
+                return result
```

Comment:
Here we don't set the result on the future because we're going to hand that off to `_run_async`